[JBIDE-23029] implemented bot test for Scaling with warning

adietish opened the pull request:
* impl'd requirements for connection, project, service
* made sure oc binary archive is NOT always downloaded
* made sure oc binary filename is platform independent

vpakan commented:
Can one of the admins verify this patch?

adietish commented:
@vpakan I'm looking for feedback so far, please dont merge it yet

adietish commented:
@jeffmaury dare to review, too?

mmalina commented:
@adietish fyi, the message from vpakan is automatic (Vlado is actually no longer in our qa team) and it is entered whenever a user not included in the jenkins job's whitelist creates a PR.

mlabuda commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/condition/ApplicationPodIsRunning.java:
```diff
@@ -55,6 +55,9 @@ public boolean test() {
 		
+		// TODO: this FLAWED: it assumes that all pods that it finds within a
+		// project are the pods for the application that a test wants to wait
+		// for
```
The wait condition is meant for a connection with default project and a single application (service). For extension of this wait condition to be usable with a project which have more pods, the constructor would have to have parameters. I would rather update javadoc of the wait condition class to be more specific about usage.

mlabuda commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/condition/NamedResourceDoesntExist.java:
```diff
@@ -0,0 +1,59 @@
+		assertNotNull(kind);
+		assertNotNull(name);
+		assertNotNull(connection);
+
```
missing assertNotNull for a project? What in case of null project? I've checked DefaultClient but it uses namespace all the time.

mlabuda commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/condition/NamedResourceDoesntExist.java:
```diff
@@ -0,0 +1,59 @@
+ * @author adietish@redhat.com
+ *
+ */
+public class NamedResourceDoesntExist extends AbstractWaitCondition {
```
Maybe do it NamedResourceExists? It's more nature for usage WaitUntil(new NamedResourceExists(...)) to wait until a resource exists or WaitWhile(new NamedResourceExists(...)) to wait while the resource still exists.

mlabuda commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/condition/OpenShiftResourceExists.java:
```diff
@@ -66,18 +68,22 @@ public ResourceExists(Resource resource, String resourceName) {
+		this(resource, new WithTextMatcher(resourceName), resourceState, null);
 	}
 	
+	public OpenShiftResourceExists(Resource resource, String resourceName, ResourceState resourceState, String projectName) {
```
add javadoc please

mlabuda commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/condition/OpenShiftResourceExists.java:
```diff
@@ -90,28 +96,41 @@ public ResourceExists(Resource resource, Matcher nameMatcher) {
+		this(resource, nameMatcher, resourceState, null);
+	}
+
+	public OpenShiftResourceExists(Resource resource, Matcher nameMatcher, ResourceState resourceState, String projectName) {
```
add javadoc please

mlabuda commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/condition/NamedResourceExist.java:
```diff
@@ -0,0 +1,64 @@
+		this(ResourceKind.PROJECT, name, null, connection);
+	}
+
+	public NamedResourceExist(String kind, String name, String project, Connection connection) {
```
Add javadoc. Its not clear enough what the kind string should be. Add reference to ResourceKind in the javadoc.

mlabuda commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/condition/ServicePodsDontExist.java:
```diff
@@ -0,0 +1,58 @@
+ * @author adietish@redhat.com
+ *
+ */
+public class ServicePodsDontExist extends AbstractWaitCondition {
```
Please make it positive: ServicePodsExist

mlabuda commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/requirement/OpenShift3ProjectRequirement.java:
```diff
@@ -0,0 +1,91 @@
+
+		String projectName = projectSpec.name();
+		Connection connection = OpenShift3ResourceUtils.getConnection(projectSpec.connectionURL());
+		if (OpenShift3ResourceUtils.hasProject(projectName, connection)) {
```
You could make negotion of condition and in body execute the code which follows condition.

adietish replied:
can you explain?

mlabuda replied:
if (!connectionDoesNotHaveProject) {

mlabuda commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/requirement/OpenShift3ProjectRequirement.java:
```diff
@@ -0,0 +1,91 @@
+		IProjectRequest request = connection.getResourceFactory().stub(ResourceKind.PROJECT_REQUEST, projectSpec.name());
+		request.setDescription(StringUtils.isEmpty(projectSpec.description())? 
+				projectSpec.name() : projectSpec.description());
+		request.setDescription(StringUtils.isEmpty(projectSpec.displayName())? 
```
there should be request.setDisplayName(...) instead of setDescription

mlabuda commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/requirement/OpenShift3ResourceUtils.java:
```diff
@@ -0,0 +1,80 @@
+	
+	private static Logger LOGGER = new Logger(OpenShift3ResourceUtils.class);
+
+	public static Connection getConnection(String connectionUrlString) {
```
Add javadoc, because it is not clear enough without description what form should the URL have. I have to look in other places in code, where its used, that it should look like "https://openshift-dev@10.1.2.2:8443".

mlabuda commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/requirement/OpenShift3ResourceUtils.java:
```diff
@@ -0,0 +1,80 @@
+	 * @return
+	 */
+	public static boolean hasProject(String name, Connection connection) {
+		return getProject(name, connection) != null;
```
getProject can throw OpenShiftToolsException. Wouldn't it be better also catch such exception and return false in that case? Because if it fails on getting a project, it like it is not there ;)

mlabuda replied:
Not directly, but nested. The request in execute method throws the exception "Unable to execute request to...". But maybe it's really rare case and we dont have to bother with it.

mlabuda commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/requirement/OpenShift3Resources.java:
```diff
@@ -0,0 +1,15 @@
+package org.jboss.tools.openshift.reddeer.requirement;
```
Please add license info.

mlabuda commented:
@adietish I've done review of changes. Please address comments where necessary.

rhopp commented:
test this

adietish commented:
@rhopp "test this"?

jkopriva commented:
@adietish we use "test this" to trigger verification job

psrna commented:
We should probably change the trigger message to something more meaningful like testPR.

mlabuda commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/requirement/OpenShift3ProjectRequirement.java:
```diff
@@ -0,0 +1,91 @@
+	@Retention(RetentionPolicy.RUNTIME)
+    @Target(ElementType.TYPE)
+    public @interface OpenShift3Project {
+		String connectionURL();
```
You missed this one.

mlabuda commented:
@adietish please address https://github.com/jbosstools/jbosstools-integration-tests/pull/1609#pullrequestreview-932050 and https://github.com/jbosstools/jbosstools-integration-tests/pull/1609#pullrequestreview-111613 and rebase PR and then I can do final tests of PR etc.

adietish commented:
addressed your comments and did quite some changes in the code. Before merging, ScalingTest should be removed and PR'd in a a separate PR (it's currently included for testing and design purposes)

mlabuda commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/requirement/OpenShiftProjectRequirement.java:
```diff
@@ -0,0 +1,134 @@
+		IProject project = OpenShiftResourceUtils.getProject(name, connection);
+		if (project == null) {
+			LOGGER.debug(NLS.bind("Project {0} doesnt exist yet in {1}, creating it.", name, connection.getHost()));
+			project = createProject(projectSpec, connection);
```
This is not OK. If default values are used ( = usage of requirement without any values in requirement annotation) and project with specified name does not exists, then it in if on line 95 it would continue to condition and on line 97 it fails (com.openshift.restclient.OpenShiftException: Exception trying to POST https://10.1.2.3:8443/oapi/v1/projectrequests response code: 422 ProjectRequest "" is invalid: metadata.name: Required value: name or generateName is required). This is because projectSpec is with no values and in create project its assumed there is value at least for project name

adietish replied:
good catch! 
it's using the wrong parameter: it should use the projectName (which is either the one specified in the projectSpec or the default), not the projectSpec which might be missing the project name

mlabuda commented:
UI is not properly updated if test class with requirements is in suite. If I expand a connection in preceding class and than in current test I try to get a project which has been just created by a requirement, it fails because connection tree item is expanded and project is not listed here. Here is minimal classes to reproduce

Suite:

> @RunWith(RedDeerSuite.class)
@SuiteClasses({
	FirstTest.class,
	SecondTest.class
})
public class OpenShiftRequirementsTestSuite {
}


First test class:
> @OpenPerspective(JBossPerspective.class)
@RequiredBasicConnection()
@RequiredProject(name=DatastoreOS3.PROJECT2)
public class FirstTest {
	@BeforeClass 
	public static void expandConnection() {
		new OpenShiftExplorerView().getOpenShift3Connection().expand();
	}
	@Test
	public void passTest() {}
} 

Second test class:
> @OpenPerspective(JBossPerspective.class)
@RequiredBasicConnection()
@RequiredProject(name=DatastoreOS3.TEST_PROJECT)
public class SecondTest {
	@InjectRequirement
	private OpenShiftProjectRequirement requiredProject;
	@InjectRequirement
	private OpenShiftConnectionRequirement requiredConnection;
	@Before
	public void getSecondProject() {
		OpenShiftExplorerView explorer = new OpenShiftExplorerView();
		explorer.open();
		explorer.getOpenShift3Connection(requiredConnection.getConnection())
			.getProject(requiredProject.getProjectName());
	}
	@Test
	public void passTest() {}
}


run test suite with following VM arguments:
>  -Dopenshift.username=openshift-dev -Dopenshift.server=https://10.1.2.2:8443 -Dopenshift.password=devel -Dusage_reporting_enabled=false -Dopenshift.authmethod=basic

adietish commented:
@mlabuda this is very weird indeed. If there's only SecondTest in the suite, everything works fine. If FirstTest and SecondTest are in the suite - while FirstTest is executed first - SecondTest fails where the project is being fetched from explorer. The project actually exists in core but the explorer seems not to get notified of it and therefore doesn't display it. Manually refreshing - while you put a brk-point into SecondTest - fixes the thing.

adietish commented:
@mlabuda I found the reason for it. We're currently not watching projects (projects are missing from the list of watched resources in WatchManager). Watching projects is an enhancement that we have filed to https://issues.jboss.org/browse/JBIDE-23184.
I could fix the issue by waiting for the resource to show up in explorer (the wait condition does a refresh). We will have to remove the workaround once the UI is getting notified of changes to the projects.

adietish commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/requirement/OpenShiftProjectRequirement.java:
```diff
@@ -0,0 +1,144 @@
+			 * 
+			 * @see WatchManager#KINDS
+			 */
+			new WaitUntil(new OpenShiftProjectExists(name, connection));
```
this wait (that refreshes the explorer/connection waits for the project to show up) fixes the issue. It should be remove once https://issues.jboss.org/browse/JBIDE-23184 is resolved

adietish commented:
@mlabuda I rebased and added a workaround/fix that's waiting & refreshing until the project gets visible in UI. Please test again and refute/accept the new PR

mlabuda commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/requirement/OpenShiftResourceUtils.java:
```diff
@@ -0,0 +1,122 @@
+			return null;
+		}
+		
+		for (IResource resource : connection.getResources(ResourceKind.PROJECT)) {
```
Replace it by safeGetResource(kind, name, null, connection).

mlabuda commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/wizard/AbstractOpenShiftApplicationWizard.java:
```diff
@@ -35,14 +36,18 @@
+		this(server, username, null);
+	}
+
+	public AbstractOpenShiftApplicationWizard(String server, String username, String project) {
```
Did not you forget to do something with the project parameter? It's completely ignored in constructor body and this constructor is not used anywhere.

mlabuda commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/wizard/v3/NewOpenShift3ApplicationWizard.java:
```diff
@@ -27,25 +29,45 @@
+public class NewOpenShift3ApplicationWizard extends AbstractOpenShiftApplicationWizard {
 
 	public NewOpenShift3ApplicationWizard() {
 		super(DatastoreOS3.SERVER, DatastoreOS3.USERNAME);
```
probably this is the place where constructor with a project parameter should be used and in open methods could be than used the project value?

mlabuda commented on tests/org.jboss.tools.openshift.ui.bot.test/src/org/jboss/tools/openshift/ui/bot/test/application/v3/advanced/ScalingTest.java:
```diff
@@ -10,111 +10,189 @@
+
+		replicationController.select();
+		scaleTo(0);
+		Matcher<String> confirmationDialogMatcher = new ShellButtonsMatcher(Arrays.asList("Yes", "No"));
```
It's really not necessary to implement such robus matcher and trying to get a shell. Please store shell name in OpenShiftLabel.Shell.STOP_DEPLOYMENT and use here new new DefaultShell(OpenShiftLabel.Shell.STOP_DEPLOYMENT));. Because waiting is default, normal, you don't have to wait for the shell, bcs. shell lookup wait for a period of lenght TimePeriod.NORMAL by default.

adietish replied:
As you said, it's superior, because it does not rely on dialog titles which are fragile. IMHO the solution is not perfect, because it relies on button labels but it is more robust because it does wait for "some dialog which has Yes/No" (to confirm scaling to 0) which is relying on the function of the upcoming dialog more than it specific outfit. Waiting for a dialog with a specific title is less robust. I dont see why we therefore should downgrade, this makes no sense to me (if you have an electric car, would you replace it by a gas driven one?).

mlabuda replied:
We have discussed on this topic in QE. We agreed this is not the right way to do stuff. You are not changing shell title. If you want to change shell title, there should be a JIRA to reflect it and not just change. And if there is a JIRA it should be reflected in changes to integration tests.

Having a shell with Yes/No is less robust because there could pop up some shell which you are not expecting which have Yes/No and it fails there.

mlabuda commented on tests/org.jboss.tools.openshift.ui.bot.test/src/org/jboss/tools/openshift/ui/bot/test/application/v3/advanced/ScalingTest.java:
```diff
@@ -10,111 +10,189 @@
+		replicationController.select();
+		scaleTo(0);
+		Matcher<String> confirmationDialogMatcher = new ShellButtonsMatcher(Arrays.asList("Yes", "No"));
+		new WaitUntil(new ShellMatchingMatcherIsAvailable(confirmationDialogMatcher));
```
Also, please remove ShellButtonsMatcher and ShellMatchingMatcherIsAvailable. There is a constructor for a DefaultShell with matcher.

adietish commented on plugins/org.jboss.tools.openshift.reddeer/src/org/jboss/tools/openshift/reddeer/condition/core/ResourceExists.java:
```diff
@@ -0,0 +1,64 @@
+ * @author adietish@redhat.com
+ *
+ */
+public class ResourceExists extends AbstractWaitCondition {
```
I dont think that this makes a lot of sense. The little code there is in both and the little code that they can share vs the additionally required checks, switches etc.

adietish commented:
@mlabuda please re-review and free for merge or refute

mlabuda commented:
Please fix ScalingTest issues. Problem is with UI elements. Even requirements create an application, UI is somehow not aware of it and thus it is failing. Most common exception:
`org.jboss.reddeer.common.exception.WaitTimeoutExpiredException: Timeout after: 10 s.: shell with title matching Matcher matching widget which text matches: "Scale Deployments" is available
	at org.jboss.reddeer.common.wait.AbstractWait.timeoutExceeded(AbstractWait.java:183)
	at org.jboss.reddeer.common.wait.AbstractWait.wait(AbstractWait.java:136)
	at org.jboss.reddeer.common.wait.AbstractWait.<init>(AbstractWait.java:101)
	at org.jboss.reddeer.common.wait.AbstractWait.<init>(AbstractWait.java:71)
	at org.jboss.reddeer.common.wait.AbstractWait.<init>(AbstractWait.java:56)
	at org.jboss.reddeer.common.wait.AbstractWait.<init>(AbstractWait.java:44)
	at org.jboss.reddeer.common.wait.WaitUntil.<init>(WaitUntil.java:35)
	at org.jboss.tools.openshift.ui.bot.test.application.v3.advanced.ScalingTest.scaleTo(ScalingTest.java:162)
	at org.jboss.tools.openshift.ui.bot.test.application.v3.advanced.ScalingTest.testScaleApplicationViaContextMenuOfService(ScalingTest.java:95)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.jboss.reddeer.junit.internal.runner.statement.RunTestMethod.evaluate(RunTestMethod.java:46)
	at org.jboss.reddeer.junit.internal.runner.statement.RunBefores.evaluate(RunBefores.java:79)
	at org.jboss.reddeer.junit.internal.runner.statement.RunIBeforeTestExtensions.evaluate(RunIBeforeTestExtensions.java:73)
	at org.jboss.reddeer.junit.internal.runner.statement.RunAfters.evaluate(RunAfters.java:68)
	at org.jboss.reddeer.junit.internal.runner.statement.RunIAfterTestExtensions.evaluate(RunIAfterTestExtensions.java:61)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.jboss.reddeer.junit.internal.runner.RequirementsRunner.runChild(RequirementsRunner.java:175)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.jboss.reddeer.junit.internal.runner.statement.RunBefores.evaluate(RunBefores.java:79)
	at org.jboss.reddeer.junit.internal.runner.statement.FulfillRequirementsStatement.evaluate(FulfillRequirementsStatement.java:45)
	at org.jboss.reddeer.junit.internal.runner.statement.RunIBeforeClassExtensions.evaluate(RunIBeforeClassExtensions.java:72)
	at org.jboss.reddeer.junit.internal.runner.statement.RunAfters.evaluate(RunAfters.java:68)
	at org.jboss.reddeer.junit.internal.runner.statement.CleanUpRequirementStatement.evaluate(CleanUpRequirementStatement.java:44)
	at org.jboss.reddeer.junit.internal.runner.statement.RunIAfterClassExtensions.evaluate(RunIAfterClassExtensions.java:59)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.jboss.reddeer.junit.internal.runner.RequirementsRunner.run(RequirementsRunner.java:156)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:678)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.jboss.reddeer.eclipse.core.RemotePluginTestRunner.main(RemotePluginTestRunner.java:68)
	at org.jboss.reddeer.eclipse.core.UITestApplication.runTests(UITestApplication.java:125)
	at org.eclipse.e4.ui.internal.workbench.swt.E4Testable$1.run(E4Testable.java:73)
	at java.lang.Thread.run(Thread.java:745)

`

mlabuda commented:
I've tried to pin down the issues. All issues are caused by 2 problems:
1) Child of project tree items are not shown correctly. Therefore there is no service visible even it is created. Could be somehow related to https://issues.jboss.org/browse/JBIDE-23184
2) Service requirement is waiting only for a replication controller. But it takes some time from moment of existing RC to the point of having right amount of application pod running. And while the test passes from service requirement to the point of scaling, where it really need application pods, it is not in correct state. Please consider using waiting for an application pod(s) in Service requirement, not only RC, so we don't have such issues next time. Or add some waiting.

mlabuda commented:
Unfortunately still it is not good. Run single ScalingTest twice, there were 2 other projects under the OpenShift connection, no test-project. Basically there was no eap service shown under a project (on screenshot, no item under test-project in OpenShift explorer view, nor any deployments in properties view, although in another IDE which I had opened I am seeing it). And it resulted with this exception:
`org.jboss.reddeer.common.exception.WaitTimeoutExpiredException: Timeout after: 60 s.: Waiting for resource Deployments matching resource name matcher a string containing "eap-app-1"
	at org.jboss.reddeer.common.wait.AbstractWait.timeoutExceeded(AbstractWait.java:183)
	at org.jboss.reddeer.common.wait.AbstractWait.wait(AbstractWait.java:136)
	at org.jboss.reddeer.common.wait.AbstractWait.<init>(AbstractWait.java:101)
	at org.jboss.reddeer.common.wait.AbstractWait.<init>(AbstractWait.java:71)
	at org.jboss.reddeer.common.wait.AbstractWait.<init>(AbstractWait.java:56)
	at org.jboss.reddeer.common.wait.WaitUntil.<init>(WaitUntil.java:47)
	at org.jboss.tools.openshift.reddeer.requirement.OpenShiftServiceRequirement.fulfill(OpenShiftServiceRequirement.java:133)
	at org.jboss.reddeer.junit.internal.requirement.Requirements.fulfill(Requirements.java:105)
	at org.jboss.reddeer.junit.internal.runner.statement.FulfillRequirementsStatement.evaluate(FulfillRequirementsStatement.java:44)
	at org.jboss.reddeer.junit.internal.runner.statement.RunIBeforeClassExtensions.evaluate(RunIBeforeClassExtensions.java:72)
	at org.jboss.reddeer.junit.internal.runner.statement.RunAfters.evaluate(RunAfters.java:68)
	at org.jboss.reddeer.junit.internal.runner.statement.CleanUpRequirementStatement.evaluate(CleanUpRequirementStatement.java:44)
	at org.jboss.reddeer.junit.internal.runner.statement.RunIAfterClassExtensions.evaluate(RunIAfterClassExtensions.java:59)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.jboss.reddeer.junit.internal.runner.RequirementsRunner.run(RequirementsRunner.java:156)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:678)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.jboss.reddeer.eclipse.core.RemotePluginTestRunner.main(RemotePluginTestRunner.java:68)
	at org.jboss.reddeer.eclipse.core.UITestApplication.runTests(UITestApplication.java:125)
	at org.eclipse.e4.ui.internal.workbench.swt.E4Testable$1.run(E4Testable.java:73)
	at java.lang.Thread.run(Thread.java:745)`

adietish commented:
@mlabuda one more set of changes. Please test them.

adietish commented:
@mlabuda with the latest set of changes I had it green 10/10 times when running ScalingTest solo and starting with non-existing test-project (test requirements create project, service, rc etc.).
Please review

adietish commented:
@rhopp I'd love if you could give this one a try, it's the required so tha people in openshift can do bot tests efficiently.

rhopp commented:
I spent quite some time with this PR and I can tell that it runs reliably and I see no obvious thing/problem/error/ with this PR which would prevent it from being merged.

adietish commented:
@rhopp great news, thanks a lot for the quick response and intensive verification!
